### ISSUE TITLE
StackOverflowError

### ISSUE THREAD

**Issue by zhen0z** [sintefmath/JutulDarcy.jl#104]
Hi Olav,

I have a problem regarding the StackOverflowError.

If I use a grid size with 30 by 30 by 7, and compute the gradient using the function:
```julia
    function objective_function(model, state, Δt, step_i, forces)
        loss = 0.0
        if step_i in steps
            sg = state.Reservoir.Saturations
            for cell in axes(sg, 2)
                loss += 0.5 * (sg[2, cell] - obs[cell]) ^ 2
            end
        end
        return loss
    end
```
everything is fine.

But when I increaes the grid size, for example, to 60 by 60 by 7, I get the following error:
```julia
JULIA: On worker 2:
StackOverflowError:
Stacktrace:
 [1] reduce_deriv!
   @ ~/.julia/packages/SparsityTracing/us9Ja/src/SparsityTracing.jl:228
 [2] reduce_deriv! (repeats 99 times)
   @ ~/.julia/packages/SparsityTracing/us9Ja/src/SparsityTracing.jl:229
Stacktrace:
  [1] (::Base.var"#989#991")(x::Task)
    @ Base ./asyncmap.jl:177
  [2] foreach(f::Base.var"#989#991", itr::Vector{Any})
    @ Base ./abstractarray.jl:3075
  [3] maptwice(wrapped_f::Function, chnl::Channel{Any}, worker_tasks::Vector{Any}, c::Vector{Vector{Any}})
    @ Base ./asyncmap.jl:177
  [4] wrap_n_exec_twice
    @ ./asyncmap.jl:153 [inlined]
  [5] #async_usemap#974
    @ ./asyncmap.jl:103 [inlined]
  [6] async_usemap
    @ ./asyncmap.jl:84 [inlined]
  [7] #asyncmap#973
    @ ./asyncmap.jl:81 [inlined]
  [8] asyncmap
    @ ./asyncmap.jl:80 [inlined]
  [9] pmap(f::Function, p::WorkerPool, c::Vector{Vector{Any}}; distributed::Bool, batch_size::Int64, on_error::Nothing, retry_delays::Vector{Any}, retry_check::Nothing)
```

How should I deal with this? Thanks.

Best regards,
Zhen

**Comment by zhen0z** [comment]
I don't think I use the setup_parameter_optimization directly. Instead, I compute the gradient like this:
'''julia

    case = JutulCase(model, dt, forces, parameters = parameters, state0 = state0)
    result = simulate_reservoir(state0, model, dt, parameters = parameters, forces = forces)

    function objective_function(model, state, Δt, step_i, forces)
        loss = 0.0
        if step_i in steps
            sg = state.Reservoir.Saturations
            for cell in axes(sg, 2)
                loss += 0.5 * (sg[2, cell] - obs[cell]) ^ 2
            end
        end
        return loss
    end

    data_domain_with_gradients = JutulDarcy.reservoir_sensitivities(case, result, objective_function)
    gradients = -data_domain_with_gradients[:permeability]

```
In this case, how do I set up use_sparsity = false?

**Comment by moyner** [comment]
I think you can do the following:
```julia
adjoint_arg = (include_state0 = include_state0, ),
data_domain_with_gradients = JutulDarcy.reservoir_sensitivities(case, result, objective_function, adjoint_arg = (use_sparsity = false, ))
```

**Comment by zhen0z** [comment]
That's great, now it works. For a permeability field with dimensinoal of 60 by 60 by 7, simulate 100 time steps with 10 days for one step, the computation time is about 70 seconds. Is it ok or should I find a way to further accelerate it? Here's the current code I use:

```julia
    function step_loss(sg, obs)
        loss = 0.0
        for cell in axes(sg, 2)
            loss += 0.5 * (sg[2, cell] - obs[cell]) ^ 2
        end
        return loss
    end
    function objective_function(model, state, Δt, step_i, forces)
        loss = 0.0
        # Compute loss if step_i is in observation steps
        if step_i in steps
            sg = state.Reservoir.Saturations
            loss += step_loss(sg, obs)
        end
        return loss
    end
    data_domain_with_gradients = JutulDarcy.reservoir_sensitivities(case, result, objective_function, adjoint_arg = (use_sparsity = false, ))
```

**Comment by moyner** [comment]
I think it sounds pretty good for that many time-steps. You may want to load HYPRE before you simulate if the case becomes big or very heterogeneous.
